### ISSUE TITLE
Remember previously seen list keys

### ISSUE THREAD

**Issue by taisph** [oliver006/redis_exporter#472]
**Describe the problem**
I'm using the exporter to monitor list key sizes for Kubernetes custom metrics to be able to scale a pod based on the size of the list key. As Redis removes a list key if it becomes empty, the exporter seems to assume the key never existed and in turn confuses the custom metrics which is told the metric doesn't exist.

Is it possible to get the exporter to remember previously seen list keys and return just return 0?

**What version of redis_exporter are you running?**
[x] 1.15.0

**Running the exporter**
```
-check-keys=lists:* -log-format=json -connection-timeout=3s
```

```
-check-single-keys=lists:my-list -log-format=json -connection-timeout=3s
```

**Comment by oliver006** [comment]
Sorry for the delayed response.
That's an interesting problem. I had a look at Prometheus docs best practices and [here](https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics) it says to avoid missing metrics.
I'm not entirely sure this 100% applies to your case but I see how it's related.
I'm a little hesitant to introduce this to the exporter ad it as keeping state of previously seen keys adds significant overhead.

As a workaround, have you considered trying to record a query as a metric (and maybe you can have it default to zerio if no data found) and then use that derived metric for your scaling?

**Comment by taisph** [comment]
> That's an interesting problem. I had a look at Prometheus docs best practices and [here](https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics) it says to avoid missing metrics.

It suggests to "export 0 ... for any time series you know may exist in advance."

> I'm a little hesitant to introduce this to the exporter ad it as keeping state of previously seen keys adds significant overhead.

Understandable.

> As a workaround, have you considered trying to record a query as a metric (and maybe you can have it default to zerio if no data found) and then use that derived metric for your scaling?

I haven't figured out a way to have the labels stick when there's no value, which means the 0 value would just be discarded by the Kubernetes custom metrics API.

**Comment by oliver006** [comment]
> > That's an interesting problem. I had a look at Prometheus docs best practices and [here](https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics) it says to avoid missing metrics.
> 
> It suggests to "export 0 ... for any time series you know may exist in advance."
> 

Correct. But these metrics are not known in advance, they're only known once a key was encountered (and that might not be until the nth scrape). Not sure where that case falls with regards to the best practices.


> > As a workaround, have you considered trying to record a query as a metric (and maybe you can have it default to zerio if no data found) and then use that derived metric for your scaling?
> 
> I haven't figured out a way to have the labels stick when there's no value, which means the 0 value would just be discarded by the Kubernetes custom metrics API.

Ugh, I see, that makes sense :-/

**Comment by taisph** [comment]
> > > That's an interesting problem. I had a look at Prometheus docs best practices and [here](https://prometheus.io/docs/practices/instrumentation/#avoid-missing-metrics) it says to avoid missing metrics.
> > 
> > It suggests to "export 0 ... for any time series you know may exist in advance."
> 
> Correct. But these metrics are not known in advance, they're only known once a key was encountered (and that might not be until the nth scrape). Not sure where that case falls with regards to the best practices.

I agree when using the scan keys option. But when explicitly listing keys using `-check-single-keys=` I would argue that the keys are expected to exist.

**Comment by oliver006** [comment]
You're right, for that case it makes more sense. I think that's something I'd be ok with adding.

**Comment by oliver006** [comment]
Closed by #483
